src/System.Memory/src/System/Buffers/ArrayBufferWriter.cs

```diff
@@ -0,0 +1,177 @@
+        }
+
+        public ReadOnlySpan<T> OutputAsSpan
+        {
+            get
+            {
+                CheckIfDisposed();
+
+                return _rentedBuffer.AsSpan(0, _index);
+            }
+        }
+
+        public int CurrentIndex
+        {
+            get
+            {
+                CheckIfDisposed();
+
+                return _index;
+            }
+        }
+
+        public int Capacity
+        {
+            get
+            {
+                CheckIfDisposed();
+
+                return _rentedBuffer.Length;
+            }
+        }
+
+        public int AvailableSpace
+        {
+            get
+            {
+                CheckIfDisposed();
+
+                return _rentedBuffer.Length - _index;
+            }
+        }
+
+        public void Clear()
+        {
+            CheckIfDisposed();
+
+            ClearHelper();
+        }
+
+        private void ClearHelper()
+        {
+            _rentedBuffer.AsSpan(0, _index).Clear();
+            _index = 0;
+        }
+
+        // Returns the rented buffer back to the pool
+        public void Dispose()
+        {
+            if (_rentedBuffer == null)
+            {
+                return;
+            }
+
+            ArrayPool<T>.Shared.Return(_rentedBuffer, clearArray: true);
+            _rentedBuffer = null;
+            _index = 0;
+        }
+
+        private void CheckIfDisposed()
+        {
+            if (_rentedBuffer == null)
+                throw new ObjectDisposedException(nameof(ArrayBufferWriter<T>));
+        }
+
+        public void Advance(int count)
+        {
+            CheckIfDisposed();
+
+            if (count < 0)
+                throw new ArgumentException(nameof(count));
+
+            if (_index > _rentedBuffer.Length - count)
+                ThrowHelper.ThrowInvalidOperationException(_rentedBuffer.Length);
+
+            _index += count;
+        }
+
+        public Memory<T> GetMemory(int sizeHint = 0)
+        {
+            CheckIfDisposed();
+
+            CheckAndResizeBuffer(sizeHint);
+            return _rentedBuffer.AsMemory(_index);
+        }
+
+        public Span<T> GetSpan(int sizeHint = 0)
+        {
+            CheckIfDisposed();
+
+            CheckAndResizeBuffer(sizeHint);
+            return _rentedBuffer.AsSpan(_index);
+        }
+
+        private void CheckAndResizeBuffer(int sizeHint)
+        {
+            if (sizeHint < 0)
+                throw new ArgumentException(nameof(sizeHint));
+
+            if (sizeHint == 0)
+            {
+                sizeHint = MinimumBufferSize;
+            }
+
+            int availableSpace = _rentedBuffer.Length - _index;
+
+            if (sizeHint > availableSpace)
+            {
+                int growBy = Math.Max(sizeHint, _rentedBuffer.Length);
+
+                int newSize = checked(_rentedBuffer.Length + growBy);
+
+                T[] oldBuffer = _rentedBuffer;
+
+                _rentedBuffer = ArrayPool<T>.Shared.Rent(newSize);
```

Comment:
This resizing code will work as one might expect, up until the limit that the ArrayPool buffers.  After that, every resize will be for the exact increase requested, rather than, say, doubling.  If we expect buffers to grow beyond a MB or two, and if common usage is to call GetSpan/Memory with relatively small values, this could get expensive quickly.

Reply:
The `CheckAndResizeBuffer` is explicitly doubling though. We grow by at least the previous rented buffer length. Asking for size of 10 after the rented buffer is already large won't grow it by just 10.
```C#
int growBy = Math.Max(sizeHint, _rentedBuffer.Length);
int newSize = checked(_rentedBuffer.Length + growBy);
_rentedBuffer = ArrayPool<T>.Shared.Rent(newSize);
```

Reply:
Ah, I missed that.  Thanks.